### PR TITLE
perf(decode): straight-loop short path + donor-gated lookahead ring + SeqSymbol repack

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Architectural rework of the sequence-section decoder hot path to donor (libzstd) shape. **-2.5% wall time on i9-9900K z000033 L-5 (725us vs 744us, 5 stable runs).**

Three architectural changes, layered:

1. **`SeqSymbol` 8-byte FSE entry** for LL/ML/OF (mirrors donor `ZSTD_seqSymbol` exactly). HUF tables keep the 12-byte `Entry` with per-state symbol byte. `FSETable` becomes `FSETableImpl<E>` generic over entry type with `FseEntry` trait.
2. **Short-block fallback path: donor `_body_*` straight loop.** Decode -> resolve offset -> execute -> next, single-pass fused. No lookahead ring overhead on hot-cache small-window frames.
3. **Long-pipeline arm: donor `_Long_body` lookahead ring restored**, with donor-equivalent gate `ddict_is_cold || (total_history > 1<<24 && offsets_long_share >= 7)`. Was previously over-engaged on 1MB frames where donor takes the straight path -- adding the 16 MB history threshold matches donor (zstd_decompress_block.c `usePrefetchDecoder` gate).

## Architecture diff vs main

| | Before | After |
|---|---|---|
| LL/ML/OF FSE entry | 12 B `Entry` (with unused `symbol`) | 8 B `SeqSymbol` (donor parity) |
| Short-block path | iterative fused with shadow_hist | straight single-pass (donor `_body`) |
| Long-pipeline gate | `cold_dict OR long_share>=7` | `cold_dict OR (hist>16MB AND long_share>=7)` |
| Long-pipeline body | 8-deep ring + prefetch | 8-deep ring + prefetch (unchanged, donor `_Long_body`) |
| DSB align | none | `.p2align 6 / nop / .p2align 5 / nop / .p2align 3` ahead of ring steady-state |

## Bench (i9-9900K, z000033 L-5, 20000 iters)

| variant | ns/iter (avg of 5) | vs main |
|---|---|---|
| main | 743,722 | -- |
| **this PR** | **724,925** | **-2.53%** |
| c_ffi (libzstd) | 445,839 | -- |

Runs: 725,930 / 724,623 / 724,994 / 724,440 / 724,639 ns (stable, std ~0.5us).

## Test plan

- [x] `cargo nextest run -p structured-zstd --features hash` 647/647
- [x] `cargo clippy -p structured-zstd --all-targets` clean
- [x] Bench on i9 with z000033 L-5 -- 5 runs stable, -2.5%
- [ ] `compare_ffi` dict groups for cold-dict path verification (gate keeps `ddict_is_cold` unconditional, hits ring on first block after `init_from_dict`)
- [ ] Full `compare_ffi` suite

## Out of scope

Remaining gap vs donor will be addressed in follow-up rounds. Identified targets ranked by flamegraph analysis: `do_offset_history_repcode` 15% self-time (donor `ZSTD_updateRep` is 5-instruction inline), FSE state advance shape (donor inlines refill inside `ZSTD_decodeSequence_bmi2` macro), literal-copy abstraction layers.

## Summary by CodeRabbit

* **Refactor**
  * Reworked sequence and entropy decoding into a unified, generic decoder design for cleaner, consistent behavior.

* **Performance**
  * Added history-aware pipeline selection and monolithic per-CPU decoders to speed common decode+execute paths and table caching.

* **Reliability**
  * Improved fallback and checkpoint/restore behavior to reduce errors and handle edge cases more robustly.

* **Tests**
  * Added unit tests covering long-pipeline gating and table/cache behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/289?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff at `sequence_section_decoder.rs:330-352` (`MIN_LONG_OFFSET_SHARE` / `HISTORY_THRESHOLD_FOR_PREFETCH` cfg gating) — verified stale. Current code at `sequence_section_decoder.rs:53-57` already has:

```rust
#[cfg(target_pointer_width = "64")]
const MIN_LONG_OFFSET_SHARE: u32 = 7;
#[cfg(not(target_pointer_width = "64"))]
const MIN_LONG_OFFSET_SHARE: u32 = 20;
const HISTORY_THRESHOLD_FOR_PREFETCH: usize = 1 << 24;
```

The two `MIN_LONG_OFFSET_SHARE` declarations are mutually exclusive via cfg — exactly one compiles per target, no duplicate-definition error. `HISTORY_THRESHOLD_FOR_PREFETCH` is unconditional but it's a plain `usize` constant that compiles on all target widths and the comparison `total_history > HISTORY_THRESHOLD_FOR_PREFETCH` is type-correct on 32-bit. No fix needed.

File: zstd/src/decoding/mod.rs [hunk 1]
```diff
@@ -55,22 +55,20 @@ pub(crate) mod prefetch;
...
+// and Sve still go through the K-generic
+// `decode_and_execute_sequences_impl` shared body until their own
+// monoliths land.
 #[cfg(target_arch = "x86_64")]
 pub(crate) mod seq_decoder_avx2;
 #[cfg(target_arch = "x86_64")]
 pub(crate) mod seq_decoder_bmi2;
+pub(crate) mod seq_decoder_scalar;
 #[cfg(target_arch = "x86_64")]
 pub(crate) mod seq_decoder_vbmi2;
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/mod.rs:86]: @coderabbitai Verified false alarm. Build is clean with strict flags on this branch tip:

- `RUSTFLAGS="-D warnings" cargo build -p structured-zstd` — no warnings
- `RUSTFLAGS="-D warnings" cargo build -p structured-zstd --tests` — no warnings
- Lint check via clippy --lib --tests — clean

The named helpers (`decode_and_execute_sequences_impl`, `run_pipelined_sequence_loop`, `decode_one_sequence_inline`, `execute_one_sequence_pipelined*`) are NOT orphaned:

1. **aarch64 production path** uses them via `CpuKernelTag::Neon` / `CpuKernelTag::Sve` dispatch arms in `decode_and_execute_sequences` (lines 180 + 190 of `sequence_section_decoder.rs`).
2. **Tests on all targets** exercise them through `#[cfg(test)]` modules.

x86_64 production goes through the per-kernel monoliths (`seq_decoder_{bmi2,avx2,vbmi2}`), but that doesn't make the shared aarch64 path dead — it's the production path for that arch. The dead_code lint correctly doesn't fire because the items have at least one reachable caller per build (either aarch64 prod or test code).

The original review said the strict-warnings build was failing on these symbols — empirically not the case.

[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/decoding/mod.rs:86]: @coderabbitai Retracting my previous reply — you were correct. My local verification ran on macOS aarch64 (M1) where the helpers ARE live (aarch64 Neon/Sve dispatch arms), so `cargo clippy` passed locally. CI on Linux x86_64 correctly surfaces the `dead_code` lint for all 11 helpers since the per-kernel monoliths bypass them entirely.

Fixed in 6ad76eef: `#[allow(dead_code)]` per-function on the orphaned helpers, with the cross-arch ones (`decode_and_execute_sequences_impl`, `run_pipelined_sequence_loop`, `decode_one_sequence_inline`, `execute_one_sequence_pipelined*`) keeping their definitions intact for aarch64 + test reuse, and the vestigial pre-R12 `_bmi2`/`_avx2`/`_vbmi2` variants flagged for cleanup in a follow-up. CI clippy gate is now green on the branch tip.

Apologies for the dismissive "false alarm" framing — the gap-closure was real, my repro was wrong.
